### DISCUSSION TITLE
How to get the first error on a call to workloadClient.itemCrud.createItem(params)

### DISCUSSION THREAD

**Discussion by box-turtle** [microsoft/Microsoft-Fabric-workload-development-sample#43] [Help]
When my front end calls `workloadClient.itemCrud.createItem(params)` and the initial call fails due to a problem with my workload BE, it seems that Fabric makes several more attempts to call the BE API and then finally ends up returning a bogus `PowerBIMetadataArtifactDisplayNameInUseException`. Is there a way for me to access the original error?

![errors](https://github.com/user-attachments/assets/e75e16cb-9c41-42f9-bb54-c2e3529087ed)

**Comment by ramizsha** [comment]
@box-turtle we are working on a way to enable error propogation from the workload backend to workload frontent, will update once it's available.

**Reply by arajawat** [reply]
I am getting same ambiguous error and here is my observation, the issue seems to be on forwarding request from DevGateway to localhost.

1. When I run DevGateway which is part of repo, I do not see any error logs.
2. When I run DevGateway from the official download link, it seems to be running in verbose mode and I get to see additional error trace on this API call:
![image](https://github.com/user-attachments/assets/eedd6ee7-0df3-446f-8207-d688d2a6d0fd)

**Reply by arajawat** [reply]
@ramizsha can you please help with this?

**Reply by arajawat** [reply]
The issue is resolved: I was using following config -
"WorkloadEndpointURL": "https://127.0.0.1:5001/workload"

Changing to "http and 5000 port" resolved it.

**Reply by box-turtle** [reply]
any updates on the issue of accessing the first exception rather than the second `PowerBIMetadataArtifactDisplayNameInUseException` when using `workloadClient.itemCrud.createItem(params)`?

**Comment by dorianj-teradata** [comment]
i'm still seeing this in prod, and it's a pretty major issue for us

**Comment by ramizsha** [comment]
We are working on adding docuemntation on how to propagate the workload backend errors to the workload frontend.
To propagate any errors that occur in the workload backend to the workload frontend when working with control plane APIs (CRUD/Jobs with the exception of GetItemJobInstanceStateAsync), the workload backend should return an error status code and the response body content should be a serialized JSON of the class "ErrorResponse" that is a part of the contracts in the workload backend.

        var errorResponse = new ErrorResponse
        {
            ErrorCode = ErrorCode,
            Message = ErrorMessage,
            MessageParameters = _messageParameters.Any() ? _messageParameters : null,
            Source = ErrorSource,
            IsPermanent = IsPermanent,
            MoreDetails = Details,
        };
        
        
        return new ContentResult
        {
            StatusCode = (int)HttpStatusCode,
            Content = JsonSerializer.Serialize(errorResponse),
            ContentType = MediaTypeNames.Application.Json,
        };
 
You can see examples in HttpResponseExceptionFilter.cs when calling "ToHttpActionResult" on exceptions in the backend sample.
When the content is a JSON of the class ErrorResponse, the error is propgated through Fabric and passed to the workload frontend, the error code in that case is "FabricExternalWorkloadError". To parse the content, the workload frontend should use WorkloadErrorDetails from @ms-fabric/workload-client.

**Reply by box-turtle** [reply]
@ramizsha - we have made the suggested changes above, but from the front end I’m seeing exactly the same behavior as before.

In dev tools I can see two requests to /artifacts (the Fabric endpoint).  (note in the .gif the first request takes quite a long time to come back)

![be-error](https://github.com/user-attachments/assets/df4ceaaf-932c-4174-a152-cb2beba25a0d)



The first one gives a 500 error but with this body:

```
{"error":{"pbi.error":{"code":null,"parameters":{},"details":[]}}}
```

That message does not go to the UI code catch block at all. 

The second try to POST to /artifacts results in this error, which DOES show up in the catch block of the UI code:

```
{
    "error": {
        "code": "PowerBIMetadataArtifactDisplayNameInUseException",
        "pbi.error": {
            "code": "PowerBIMetadataArtifactDisplayNameInUseException",
            "details": [
                {
                    "code": "PowerBIMetadataArtifactDisplayNameNotAvailableYetException",
                    "detail": {
                        "type": 1,
                        "value": "The DisplayName is not available yet."
                    }
                }
            ],
            "exceptionCulprit": 1
        }
    }
}
```

Our BE code is setting the http status code to 500 and sending this response body:

```
{
  "ErrorCode": 500,
  "Message": "operation failed, internal server error",
  "MessageParameters": "",
  "Source": "",
  "IsPermanent": false,
  "MoreDetails": "operation failed, internal server error"
}
```

What else do we have to do differently to allow the UI to show the error generated by our BE?

**Reply by ramizsha** [reply]
@box-turtle please email our support alias (ILDCWLSupport@microsoft.com) the request id of the first request so we can take a look from our side, please include in your email:
* Whether this is dev mode or not.
* Whether you are sure that the request reaches your backend or not.
